### ISSUE TITLE
Port twisted.web.resource to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6078]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6078|
|Type|enhancement|
|Created|2012-10-08 11:04:09Z|
|Branch|https://github.com/twisted/twisted/tree/resource-py3-6078-2|

The basis of the resource publishing model for the Twisted Web server.

<details><summary>Searchable metadata</summary>

```
trac-id__6078 6078
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_resource_py3_6078_2 branches/resource-py3-6078-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1349694249000000 1349694249000000
changetime__1350258602000000 1350258602000000
version__None None
owner__exarkun exarkun
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [36108]) Branching to 'resource-py3-6078'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Basically done, I think, except:

  1. depends on `prefixedMethodNames` ([#6100](https://github.com/twisted/twisted/issues/6100))
  1. I threw some `encode('ascii')` and `decode('ascii')` in.  They get the job done in the naive case, but maybe they merit some additional error handling code?  Or something?  Or maybe 'latin-1' is a better charset to use.  Also, maybe they suck for the Python 2 case - time for another string type/compat helper?

Maybe other stuff is terrible about this too, but those stand out to me.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst|
|-|-|

Still depends on [#6100](https://github.com/twisted/twisted/issues/6100) being resolved first.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set owner to @exarkun|
|-|-|

Nice to see extra tests.

  1. I'm fine with HTTP methods being ASCII only (latin-1 doesn't make sense to me; I suppose it's possible RFC will be expanded to be UTF-8 someday, but not that). For the decode case, `nativeString` is what you want to use, I believe; in particular since on Python 2 this is a slightly more expensive way of doing what the current trunk code does, so no change or potential for new errors. For the encode... we could create `nativeStringToBytes`? I suppose if we find another instance where this is necessary we should create it, but for now I'd say just leave it as is.
  1. `Resource` and `IResource` could use some addition and clarification of the types of parameters. E.g. `IResource` doesn't mention types at all, and `Resource` says string when it means bytes.
  1. `Resource.getChild` code example should indicate bytes, likewise `putChild`.
  1. Add the byteness of path elements to the porting howto.

Fix the above, then merge (after [#6100](https://github.com/twisted/twisted/issues/6100) I guess).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [36145]) Fixes to string types used by ErrorPage-related code.

refs [#6078](https://github.com/twisted/twisted/issues/6078)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [36147]) Branching to 'resource-py3-6078-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

> I'm fine with HTTP methods being ASCII only (latin-1 doesn't make sense to me; I suppose it's possible RFC will be expanded to be UTF-8 someday, but not that). For the decode case, nativeString is what you want to use, I believe; in particular since on Python 2 this is a slightly more expensive way of doing what the current trunk code does, so no change or potential for new errors. For the encode... we could create nativeStringToBytes? I suppose if we find another instance where this is necessary we should create it, but for now I'd say just leave it as is.

Switched to `nativeString` and added a comment for the encode case in r36142.

> Resource and IResource could use some addition and clarification of the types of parameters. E.g. IResource doesn't mention types at all, and Resource says string when it means bytes.
> Resource.getChild code example should indicate bytes, likewise putChild.

Fixed in r36143.

> Add the byteness of path elements to the porting howto.

Did that in r36144.

Also fixed some `ErrorPage` stuff in r36145 and r36146.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [36151]) Merge resource-py3-6078-2

Author: exarkun
Reviewer: itamarst
Fixes: [#6078](https://github.com/twisted/twisted/issues/6078)

Port `twisted.web.error` and `twisted.web.resource` to Python 3.  Also add some
unit tests for `twisted.web.resource.Resource` which were previously missing.
